gh-137716: TracebackException to handle messages with punctuation

krisztian-gajdar opened the pull request:
Added extra handling for error messages with punctuation to avoid double punctuations and eliminated code duplication


<!-- gh-issue-number: gh-137716 -->
* Issue: gh-137716
<!-- /gh-issue-number -->

krisztian-gajdar committed:
fix(traceback): TracebackException to handle error messages with punctuations

docs: add news entry for TracebackException punctuation fix

style: remove trailing whitespace in traceback.py

Update 2025-08-24-11-42-38.gh-issue-137716.7-Mtj-.rst

Update 2025-08-24-11-42-38.gh-issue-137716.7-Mtj-.rst

refactor: improve suggestion message computation and punctuation handling

style: remove trailing whitespace in traceback.py

gh-137986: Fix and improve the csv functions docstrings (GH-137987)

The csv.register_dialect() docstring no longer imply that it returns a
dialect.
All functions have now signatures.

Co-authored-by: maurycy <5383+maurycy@users.noreply.github.com>

gh-135261: bring back CI job for testing OpenSSL 1.1.1w (#135262)

This partially reverts commit d83e30caddcbf9482273743d287577517ec735b7
by bringing back the CI job for testing OpenSSL 1.1.1w. Despite this
version being upstream EOL, the rationale for keeping it as follows:

- It most resembles other 1.1.1-work-a-like ssl APIs supported by important vendors.
- Python officially requires OpenSSL 1.1.1 or later, although OpenSSL 3.0 or later
  is recommended for cryptographic modules. Since changing the build requirements
  requires a transition period, we need to keep testing the allowed versions.
- The code base still contains calls to OpenSSL functions that are deprecated since
  OpenSSL 3.0 as well as `ifdef` blocks constrained to OpenSSL 1.1.1.

picnixz reviewed (dismissed):
There seems to be too many changes. Please revert unrelated changes and code refactoring.

picnixz commented on Lib/traceback.py:
```diff
@@ -1081,50 +1081,30 @@ def __init__(self, exc_type, exc_value, exc_traceback, *, limit=None,
-            wrong_name = getattr(exc_value, "name_from", None)
-            suggestion = _compute_suggestion_error(exc_value, exc_traceback, wrong_name)
-            if suggestion:
-                self._str += f". Did you mean: '{suggestion}'?"
```
Here, just do:

```py
self._str = self._str.removesuffix('.') + f". Did you mean: '{suggestion}'?"
```

krisztian-gajdar replied:
I'd have to do that in 4 different places. My change only includes refactoring to avoid this while also reducing complexity. I was following the along the lines of `_compute_suggestion_error` and separating the message building part into a separate function as well. Please reconsider. 🙏

picnixz commented on Lib/traceback.py:
```diff
@@ -1081,50 +1081,30 @@ def __init__(self, exc_type, exc_value, exc_traceback, *, limit=None,
+                self._is_syntax_error = True
+                self._exc_metadata = getattr(exc_value, "_metadata", None)
+            elif suggestion := _compute_suggestion_message(exc_type, exc_value, exc_traceback):
+                punctuation = "" if self._str.rstrip().endswith(('.', '!')) else "."
```
Don't rstrip() the entire string (I think if someone puts spaces at the end, they want those spaces; it may not be optimal but I don't think we need to necessarily rstrip the message). We can do the following as well (I'm feeling that the line is very long, please wrap them under 80 chars as per PEP-8):

```py
if self._str.endswith(('.', '?', '!', '...')):
    punctuation = ''
else:
    punctuation = '.'
```

picnixz commented on Lib/traceback.py:
```diff
@@ -1732,6 +1710,26 @@ def _compute_suggestion_error(exc_value, tb, wrong_name):
     return suggestion
 
 
+def _compute_suggestion_message(exc_type, exc_value, exc_traceback) -> str | None:
```
```suggestion
def _compute_suggestion_message(exc_type, exc_value, exc_traceback):
```

We don't add typings in the standard library usually. It's left to typeshed.

picnixz commented on Lib/traceback.py:
```diff
@@ -1732,6 +1710,26 @@ def _compute_suggestion_error(exc_value, tb, wrong_name):
+        sys.flags.no_site and \
+        getattr(exc_value, "name", None) not in sys.stdlib_module_names:
+            return ("Site initialization is disabled, did you forget to "
+                "add the site-packages directory to sys.path?")
```
```suggestion
    if (
        issubclass(exc_type, ModuleNotFoundError)
        and sys.flags.no_site
        and getattr(exc_value, "name", None) not in sys.stdlib_module_names
    ):
       return ("Site initialization is disabled, did you forget to "
               "add the site-packages directory to sys.path?")
```

picnixz commented on Lib/traceback.py:
```diff
@@ -1732,6 +1710,26 @@ def _compute_suggestion_error(exc_value, tb, wrong_name):
+                "add the site-packages directory to sys.path?")
+    if issubclass(exc_type, (ImportError, NameError, AttributeError)):
+        name_attribute = "name_from" if issubclass(exc_type, ImportError) else "name"
+        wrong_name = getattr(exc_value, name_attribute, None)
```
```py
    if issubclass(exc_type, (ImportError, NameError, AttributeError)):
        if issubclass(exc_type, ImportError):
            wrong_name = getattr(exc_value, "name_from", None) 
        else:
            wrong_name = getattr(exc_value, "name", None) 
```

This one is up to you. Just don't exceed 80 chars.

picnixz commented on Lib/traceback.py:
```diff
@@ -1732,6 +1710,26 @@ def _compute_suggestion_error(exc_value, tb, wrong_name):
+                if suggestion_error:
+                    return f"{suggestion_text} Or did you forget to import '{wrong_name}'?"
+                return f"Did you forget to import '{wrong_name}'?"
+            return suggestion_text
```
I think the following could be a bit clearer:

```suggestion
            other_name = _compute_suggestion_error(
                exc_value, exc_traceback, wrong_name
            )
            maybe_builtin_import = (
                issubclass(exc_type, NameError)
                and wrong_name in sys.stdlib_module_names
            )
            if not other_name:
                if maybe_builtin_import:
                    return f"Did you forget to import '{wrong_name}'?"
                return None
            text = f"Did you mean: '{other_name}'?"
            if maybe_builtin_import:
                return f"{text} Or did you forget to import '{wrong_name}'?"
            return text
```

picnixz reviewed (commented):
Ok for the refactoring (no need for additional tests). However here are some suggestions for readability:

krisztian-gajdar committed:
revert unrelated changes

krisztian-gajdar committed:
revert unrelated changes

AA-Turner commented on Lib/traceback.py:
```diff
@@ -1081,50 +1081,33 @@ def __init__(self, exc_type, exc_value, exc_traceback, *, limit=None,
             self.exc_type_qualname = exc_type.__qualname__
             self.exc_type_module = exc_type.__module__
+            if issubclass(exc_type, SyntaxError):
+                # Handle SyntaxError's specially
```
Bad grammar (apostrophe). Does it make sense to fold this into the new function? I'd remove the compute prefix though, style in Python is to avoid things like get_, set_, etc.

AA-Turner commented on Lib/traceback.py:
```diff
@@ -1081,50 +1081,33 @@ def __init__(self, exc_type, exc_value, exc_traceback, *, limit=None,
+                self._is_syntax_error = True
+                self._exc_metadata = getattr(exc_value, "_metadata", None)
+            elif suggestion := _compute_suggestion_message(exc_type, exc_value, exc_traceback):
+                if self._str.endswith(('.', '?', '!', '...')):
```
Ellipsis is covered by the full stop.
```suggestion
                if self._str.endswith(('.', '?', '!')):
```

AA-Turner commented on Lib/traceback.py:
```diff
@@ -1384,7 +1367,6 @@ def _find_keyword_typos(self):
                 self.msg = f"invalid syntax. Did you mean '{suggestion}'?"
                 return
 
-
```
Unrelated change, please revert this and the others.

krisztian-gajdar committed:
refactor: rename and simplify error suggestion handling

Rename `_compute_suggestion_message` to  `_suggestion_message`
and simplify the punctuation check. reverted whitespace changes.

krisztian-gajdar committed:
test: add test_exception_punctuation_handling_with_suggestions

krisztian-gajdar committed:
test: add test_exception_punctuation_handling_with_suggestions

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -4878,6 +4878,28 @@ class PurePythonSuggestionFormattingTests(
+                )
+                expected = f"NameError: Error{punctuation} Did you forget to import 'time'?"
+                self.assertEqual(result_lines[0], expected)
+
```
```suggestion
                self.assertEqual(result_lines[0], expected)
```

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -4878,6 +4878,28 @@ class PurePythonSuggestionFormattingTests(
     traceback printing in traceback.py.
     """
 
+    def test_exception_punctuation_handling_with_suggestions(self):
```
Is it really the place where to put this test? or should it be somewhere else? (sorry but the file is 5k lines long so it's a bit hard to know...)

picnixz replied:
Can you add more tests where we don't hit the NameError + `maybe_builtin_import` path by changing `name=` to something that is not in `sys.stdlib_module_names`. (For instance `foo123`).

picnixz commented on Lib/traceback.py:
```diff
@@ -1732,6 +1715,38 @@ def _compute_suggestion_error(exc_value, tb, wrong_name):
+            other_name = _compute_suggestion_error(
+                exc_value, exc_traceback, wrong_name
+            )
+            maybe_builtin_import = (
```
I just remembered that, in addition to `sys.stdlib_module_names`, we also have `sys.builtin_module_names`, so let's rename this variable `maybe_stdlib_import` instead of `maybe_builtin_import`.

krisztian-gajdar committed:
refactor: review items

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -3802,6 +3802,26 @@ def test_traceback_header(self):
+            try:
+                raise NameError(message, name=name)
+            except Exception as e:
+                return traceback.TracebackException.from_exception(e)._str
```
Can't you use the `list(exc.format())` one instead? it's better than accessing a private attribute.

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -3802,6 +3802,26 @@ def test_traceback_header(self):
+        def raise_mssage(message, name):
+            try:
+                raise NameError(message, name=name)
+            except Exception as e:
```
```suggestion
            except NameError as e:
```

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -3802,6 +3802,26 @@ def test_traceback_header(self):
+    def test_exception_punctuation_handling_with_suggestions(self):
+        def raise_mssage(message, name):
+            try:
+                raise NameError(message, name=name)
```
Please test something that is not a NameError as well.

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -3802,6 +3802,26 @@ def test_traceback_header(self):
+                self.assertEqual(messsage, expected)
+
+
     @requires_debug_ranges()
```
```suggestion
                self.assertEqual(messsage, expected)

    @requires_debug_ranges()
```

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -3802,6 +3802,26 @@ def test_traceback_header(self):
+            ("Error.", "time", "Error. Did you forget to import 'time'?"),
+            ("Error?", "time", "Error? Did you forget to import 'time'?"),
+            ("Error!", "time", "Error! Did you forget to import 'time'?"),
+            ("Error", "time", "Error. Did you forget to import 'time'?"),
```
Please add a test where there is the "Or did you forget [...]" part.

krisztian-gajdar replied:
It is covered here: https://github.com/krisztian-gajdar/cpython/blob/traceback-to-handle-error-msgs-with-period/Lib/idlelib/idle_test/test_run.py#L42
I think the new tests should focus on the punctuation handling, as other functionalities are already covered.
But I added tests for ImportError as well

picnixz replied:
I would forget about the fact that it's in the IDLE module. We sometimes have multiple tests covered here and there. This PR refactors code, so it's still good to have coverage even if it's covered twice. Having it covered where it *should* be covered directly would be better. In IDLE, it's an indirect coverage as it doesn't explicitly tests traceback itself IMO.

krisztian-gajdar replied:
@picnixz Understood, added new test cases, let me know if it is sufficient

krisztian-gajdar committed:
test: extend error suggestion punctuation handling tests

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -3802,6 +3802,45 @@ def test_traceback_header(self):
+                return list(exc.format())[-1]
+
+        test_cases = [
+            ("a.", "time", "NameError: a. Did you forget to import 'time'?\n"),
```
Please duplicate the NameError from the IDLE tests here where we test "Or did you forget?".

krisztian-gajdar replied:
done

krisztian-gajdar committed:
test: new traceback test cases

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -3802,6 +3802,58 @@ def test_traceback_header(self):
         self.assertEqual(list(exc.format()), ["Exception: haven\n"])
 
+    def test_name_error_punctuation_with_suggestions(self):
+        def raise_mssage(message, name, name_from=None):
```
Let's refactor this as a separate method helper:

```py
@staticmethod
def format_error(exc_type, exc_message, *args, **kwargs):
    try:
        raise exc_type(exc_message, *args, **kwargs)
    except exc_type as e:
        exc = traceback.TracebackException.from_exception(e)
        return list(exc.format())[-1]
```

Then use this helper directly in your tests but you can have a local helper as well:

```py
# in the test itself:
def format_error(message, name):
    return self.format_error(NameError, message, name=name)
```

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -3802,6 +3802,58 @@ def test_traceback_header(self):
+        ]
+        for message, name, expected in test_cases:
+            with self.subTest(message=message):
+                messsage = raise_mssage(message, name)
```
And here you would use `message = format_error(message, name)`.

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -3802,6 +3802,58 @@ def test_traceback_header(self):
+            self.assertEqual(messsage, expected_message)
+
+    def test_import_error_punctuation_handling_with_suggestions(self):
+        def raise_mssage(message):
```
Here you would define:

```py
def format_error(message):
    self.format_error(ImportError, message, name="math", name_from="sinq")
```

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -3802,6 +3802,58 @@ def test_traceback_header(self):
+        ]
+        for message, expected in test_cases:
+            with self.subTest(message=message):
+                messsage = raise_mssage(message)
```
And here you would have `message = format_error(message)`

picnixz commented on Lib/test/test_traceback.py:
```diff
@@ -3802,6 +3802,58 @@ def test_traceback_header(self):
+                messsage = raise_mssage(message, name)
+                self.assertEqual(messsage, expected)
+
+        with self.subTest("combined suggestion"):
```
```suggestion
        with self.subTest("stdlib module import suggestion"):
```

picnixz reviewed (commented):
Some refactoring suggestions.

krisztian-gajdar committed:
refactor traceback tests

krisztian-gajdar committed:
Merge branch 'main' into traceback-to-handle-error-msgs-with-period

krisztian-gajdar committed:
fix: remove suggestion prompts from exception error messages

Remove "Did you mean" suggestions and import prompts from NameError and
AttributeError exception handling in traceback formatting. This eliminates
automatic suggestion text that was being appended to error messages.

krisztian-gajdar committed:
Merge branch 'main' into traceback-to-handle-error-msgs-with-period

krisztian-gajdar committed:
Merge branch 'main' into traceback-to-handle-error-msgs-with-period

krisztian-gajdar committed:
Merge branch 'main' into traceback-to-handle-error-msgs-with-period

krisztian-gajdar committed:
Merge branch 'main' into traceback-to-handle-error-msgs-with-period

krisztian-gajdar committed:
Merge branch 'main' into traceback-to-handle-error-msgs-with-period

krisztian-gajdar committed:
refactor(traceback): consolidate ModuleNotFoundError handling and fix suggestion formatting

Move ModuleNotFoundError stdlib module detection into the unified
suggestion handling block, removing duplicate exception handling code.

Also fix the site initialization disabled message to include virtual
environment hint and add proper handling for non-ASCII suggestion names
in the _suggestion_message function.

krisztian-gajdar committed:
Merge branch 'main' into traceback-to-handle-error-msgs-with-period
